Add the ability to import checklists from Trello

leosunmo opened the pull request:
#### Summary
Adds the ability to import checklists to Focalboard when importing from Trello. It's pretty simple code, but it's my first attempt at TypeScript, so appreciate any feedback. I'm not 100% sure of my `undefined` logic etc. But it does work :shrug: 

This is kind of related to https://github.com/mattermost/focalboard/issues/53 but there's no specific issue relating to this. I wrote it for my own needs and figured I should push it upstream.

I started looking at importing images as well, but they require auth now if the Trello board is private. That is a different kettle of fish, so decided to leave that out of this PR.

mattermod commented:
Hello @leosunmo,

Thanks for your pull request! A Core Committer will review your pull request soon. For code contributions, you can learn more about the review process [here](https://developers.mattermost.com/contribute/getting-started/code-review/).

Per the [Mattermost Contribution Guide](https://developers.mattermost.com/contribute/getting-started/), we need to add you to the list of [approved contributors](https://docs.google.com/spreadsheets/d/1NTCeG-iL_VS9bFqtmHSfwETo5f-8MQ7oMDE5IUYJi_Y/pubhtml?gid=0&single=true) for the Mattermost project.

**Please help complete the Mattermost [contribution license agreement](http://www.mattermost.org/mattermost-contributor-agreement/)?**
Once you have signed the CLA, please comment with `/check-cla` and confirm that the CLA check is green.

This is a standard procedure for many open source projects.

Please let us know if you have any questions.

We are very happy to have you join our growing community! If you're not yet a member, please consider joining our [Contributors community channel](https://pre-release.mattermost.com/core/channels/tickets) to meet other contributors and discuss new opportunities with the core team.

leosunmo commented:
/check-cla

leosunmo commented:
/check-cla

chenilim commented on import/trello/importTrello.ts:
```diff
@@ -135,6 +136,29 @@ function convert(input: Trello): Block[] {
+                    lookup.checkItems.forEach(trelloCheckBox=> {
+                        const checkBlock = createCheckboxBlock()
+                        checkBlock.title = trelloCheckBox.name
+                        if (trelloCheckBox.state == 'complete') {
```
The eslint setup for the import scripts may be out of date from the webapp, but in general we should use ===, i.e. `if (trelloCheckBox.state === 'complete') {`.

chenilim commented on import/trello/importTrello.ts:
```diff
@@ -135,6 +136,29 @@ function convert(input: Trello): Block[] {
+        if (card.idChecklists && card.idChecklists.length) {
+            card.idChecklists.forEach(checklistID => {
+                const lookup = input.checklists.find(e => e.id === checklistID)
+                if (lookup !== undefined) {
```
In general, we can check if a variable is "tuthy" i.e. `if (lookup) {`, which will handle undefined, null, and other cases that can lead to a null reference exception.

chenilim commented on import/trello/importTrello.ts:
```diff
@@ -135,6 +136,29 @@ function convert(input: Trello): Block[] {
         }
+
+        // Add Checklists
+        if (card.idChecklists && card.idChecklists.length) {
```
This is minor, and may be stylistic, but in general lengths should be checked as > 0, i.e. ` if (card.idChecklists && card.idChecklists.length > 0) {`.

leosunmo replied:
That was another thing I was wondering about. I originally used `> 0` (habit from writing Go) but kept seeing people doing what I ended up doing :shrug: so I wasn't sure if there was some hidden syntactic thing or just a style thing. I definitely prefer the clarity of `> 0`.

chenilim commented on import/trello/importTrello.ts:
```diff
@@ -9,7 +9,8 @@ import {IPropertyOption, IPropertyTemplate, createBoard} from '../../webapp/src/
 import {createCard} from '../../webapp/src/blocks/card'
 import {createTextBlock} from '../../webapp/src/blocks/textBlock'
-import {Trello} from './trello'
+import {createCheckboxBlock, CheckboxBlock} from '../../webapp/src/blocks/checkboxBlock'
```
ESlint called out that `CheckboxBlock` and `ChecklistElement` aren't used, and should not be imported.

leosunmo replied:
Yeah my editor says the same, not sure why they're still there to be honest, must've had a language server error when I tidied and committed. Had to set up the TypeScript/JavaScript environment from scratch for this.

chenilim reviewed (changes requested):
Thanks @leosunmo for your contribution! This is a great addition to the import script! Logic looks good, just a few minor linter errors - the import scripts aren't part of the main CI / lint check.

leosunmo commented:
Take a look at my pushed changed. Pretty sure I addressed all your concerns. Really appreciate the language/style pointers, as I said, I've stayed away from JS/TS :p

chenilim reviewed (approved):
Thanks @leosunmo! LGTM 👍

emilyacook commented:
Thanks for participating in Hacktoberfest! You can claim your sticker set here: https://get.printfection.com/hacktober21/4144583267 @leosunmo
